scan-payment/src/main/java/com/getbouncer/scan/payment/ml/ExpiryDetect.kt

```diff
@@ -0,0 +1,145 @@
+
+    data class Expiry(val month: Int, val year: Int) {
+        override fun toString() = formatExpiry(
+            day = null,
+            month = month.toString(),
+            year = year.toString()
+        )
+    }
+
+    private data class Digit(val digit: Int, val confidence: Float)
+
+    override val name: String = Factory.NAME
+
+    override fun buildEmptyMLOutput() = arrayOf(arrayOf(Array(NUM_PREDICTIONS) { FloatArray(NUM_CLASS) }))
+
+    override fun interpretMLOutput(data: Input, mlOutput: Array<Array<Array<FloatArray>>>): Prediction {
+        val output = mlOutput[0][0].mapNotNull {
+            it.indexOfMax()?.let { maxIndex ->
+                Digit(maxIndex, it[maxIndex])
+            }
+        }
+
+        val (newDigits, confidence) = output.map {
+            Pair(it.digit, it.confidence)
+        }.unzip()
+
+        val digits = greedyNonMaxSuppression(
+            newDigits.toTypedArray(),
+            confidence.toFloatArray(),
+            BACKGROUND_CLASS
+        ).filter { it != BACKGROUND_CLASS }
+
+        return if (digits.size == 4 || (digits.size == 5 && digits[2] == 1)) {
+            // process if we get exactly 4 digits, OR it's five digits and the middle prediction
+            // is a 1 - this is because we sometimes mistake '/' for '1'
+            val month = digits[0] * 10 + digits[1]
+            val year = 2000 + digits[digits.size - 2] * 10 + digits[digits.size - 1]
+            if (isValidMonth(month)) {
+                Prediction(Expiry(month, year))
+            } else {
+                Prediction(null)
+            }
+        } else {
+            Prediction(null)
+        }
+    }
+
+    override fun transformData(data: Input): ByteBuffer {
+        val targetAspectRatio = ASPECT_RATIO
+        val scaledRect = data.expiryBox.scaled(data.image.size())
+        val scaledExpRectNewHeight = scaledRect.width() * targetAspectRatio
+
+        val rect = Rect(
+            scaledRect.left.roundToInt(),
+            (scaledRect.centerY() - scaledExpRectNewHeight / 2).roundToInt(),
+            scaledRect.right.roundToInt(),
+            (scaledRect.centerY() + scaledExpRectNewHeight / 2).roundToInt()
+        )
+
+        return data.image
+            .crop(rect)
+            .scale(TRAINED_IMAGE_SIZE)
+            .toRGBByteBuffer()
+    }
+
+    override fun executeInference(
+        tfInterpreter: Interpreter,
+        data: ByteBuffer,
+        mlOutput: Array<Array<Array<FloatArray>>>
+    ) = tfInterpreter.run(data, mlOutput)
+
+    /**
+     * A factory for creating instances of the [ScreenDetect]. This downloads the model from the
+     * web. If unable to download from the web, this will throw a [FileNotFoundException].
+     */
+    class Factory(context: Context, loader: Loader) : TFLAnalyzerFactory<ExpiryDetect>(loader) {
+        companion object {
+            private const val USE_GPU = false
+            private const val NUM_THREADS = 2
```

Comment:
I've been moving these to parameters on the factory constructor so that we can decide at runtime how many threads we want. this was useful for testing loops with high thread counts and low parallelism.